lib/azure/armrest/armrest_manager.rb

```diff
@@ -209,14 +209,71 @@ def providers
 
       # Returns information about the specific provider +namespace+.
       #
-      def provider_info(namespace)
-        url = url_with_api_version(@base_url, 'providers', namespace)
+      def provider_info(provider)
+        url = url_with_api_version(@base_url, 'providers', provider)
         response = rest_get(url)
         JSON.parse(response.body)
       end
 
       alias geo_locations provider_info
 
+      # Returns a list of all locations for all resource types of the given
+      # +provider+. If you do not specify a provider, then the locations for
+      # all providers will be returned.
+      #
+      # If you need individual details on a per-provider basis, use the
+      # provider_info method instead.
+      #--
+      # TODO: I think a 7-day cache may be wise here as the overall list of
+      # locations is unlikely to change from day to day.
+      #
+      def locations(provider = nil)
+        array = []
+
+        if provider
+          info = provider_info(provider)
+          info['resourceTypes'].map{ |rt| array << rt['locations'] }
+        else
+          threads = []
+
+          providers.each do |provider_hash|
+            provider = provider_hash['namespace']
+            threads << Thread.new{
+              info = provider_info(provider)
+              info['resourceTypes'].map{ |rt| array << rt['locations'] }
+            }
+          end
+
+          threads.each(&:join)
+        end
+
+        array.flatten!
+        array.uniq!
+        array.delete("") # Seems to pickup blank elements for some reason.
+
+        array
+      end
+
+      # Returns an array of publishers for the given +region+ and +provider+,
+      # or "Microsoft.Compute" if no provider is given.
+      #
+      # Examples:
+      #
+      #   arm.publishers('eastus')
+      #   arm.publishers('eastus', 'Microsoft.ClassicCompute')
+      #
+      def publishers(region, provider = 'Microsoft.Compute', subscription_id = @subscription_id)
+        @api_version = '2015-06-15'
```

Comment:
Oh, you can set that either in the configure method or on a per object basis either in the constructor or after the fact on an object. You don't want to set that on a per-gem basis because of the unstable nature of the API at the moment because not every version string works for every method (yay, microsoft). In some cases I explicitly set it within a method because I -happen- to know that the default api_version won't work.